### PR TITLE
SimulatorFullyImplicitBlackoilOutput: add FLOWS to the output.

### PR DESCRIPTION
**Pull request by dr-robertk**
If the keyword FLOWS was added to the deck then this patch will add the necessary output of data. 
Most of the patch was provided by @osae.

### CODE REVIEW

File: opm/autodiff/SimulatorFullyImplicitBlackoilOutput.hpp [hunk 1]
```diff
@@ -392,10 +400,69 @@ namespace Opm
...
+        for (size_t face_idx = 0; face_idx < num_faces; ++face_idx) {
+          auto c1 = fc(face_idx, 0);
+          auto c2 = fc(face_idx, 1);
+
+          if (c1 == -1 || c2 == -1)
+            continue; // face on grid boundary
+
+          int gc1 = gc ? gc[c1] : c1;
+          int gc2 = gc ? gc[c2] : c2;
+          int gdiff = std::abs(gc2-gc1);
```
Comment (blattms) [inline opm/autodiff/SimulatorFullyImplicitBlackoilOutput.hpp]: But why bother as we need the global values anyway.

**Comment by atgeirr** [comment]
After reading the code I am convinced that there is one bug here: the non-cartesian connections are not handled properly. It is incorrect to associate one N+ entry with each cell, since across a fault one cell can have many non-cartesian connections. The size of this field should instead be equal to the number of non-cartesian connections, the easiest way to access these connections is through DerivedGeology::nonCartesianConnections().

**Comment by joakim-hove** [comment]
>It is incorrect to associate one N+ entry with each cell, since across a fault one cell can have many non-cartesian connections [....]

Well - what is *correct* can probably be discussed at length, but if the `FLOW` property should go to the restart files in a ~Eclipse compatible way there should be one `N+` entry for each cell, and the fault case must be handled differently - resembling the Eclipse NNC treatment we already have. The output layer is not yet ready handle the fault specific/NNC part of this, but that should be a limited task.

**Comment by atgeirr** [comment]
> Well - what is correct can probably be discussed at length, but if the FLOW property should go to the restart files in a ~Eclipse compatible way there should be one N+ entry for each cell

Then there is something I did not understand. Will investigate this before I open my mouth again...

**Comment by atgeirr** [comment]
Also: must be rebased to be mergeable.

**Comment by joakim-hove** [comment]
> Then there is something I did not understand. Will investigate this before I open my mouth again...

Or - heaven forbid; the misunderstanding could be on my side ...

**Comment by atgeirr** [comment]
Ok, I checked some eclipse output files, and for FLROILN+ (which should behave just as the currently considered fields, just with reservoir not surface volumes) we have 180 elements. This case has 4000 cells (and FLROILI+ has 4000 elements), and 180 NNCs, so my conclusion is that the current implementation is not correct.

**Comment by dr-robertk** [comment]
@atgeirr: Thanks for the thorough review. I will revise this and rebase on the current master. Might take a while.

**Comment by atgeirr** [comment]
I am still interested in seeing this revised and completed, so: should we keep this open?

**Comment by atgeirr** [comment]
@dr-robertk I would like to move forward on this, would you like to do it? Otherwise I'd like to start on it quite soon myself.

**Comment by dr-robertk** [comment]
@atgeirr: This is probably a stale PR and should be closed.

**Comment by matteodg** [comment]
> @atgeirr: This is probably a stale PR and should be closed.

Any chance to have this PR updated to the current code and working? It would be great to have the flows properties in output. Thanks in advance for any comment/hint on this.

**Comment by atgeirr** [comment]
> Any chance to have this PR updated to the current code and working? It would be great to have the flows properties in output. Thanks in advance for any comment/hint on this.

It is on our agenda (again)! But other improvements (to multisegment wells for example) have higher priority right now, so I am afraid I cannot promise it by a particular date.

**Comment by bska** [comment]
I'm going to close this now.  The code-base has changed too much since this was proposed so we're better off starting from scratch to add the flux outputs.

**Pull request closed without merging**